Compilation failure in `domain_state.h` for librairies with OCaml ≥ 4.10

Ha!

I think I’ve found a bug in OCaml! A real one ;-)
The bug was introduced in commit fc6f0284922621fc8717033a8d73b1134f1506ac and affects OCaml ≥ 4.10.
Pinging the contributors to this file: @kayceesrk, @Engil, and @dra27.


In Opam, we’re transitionning to [OCaml 4.10 for bootstrap][41000-bootstrap], and we’ve
had this issue in [Opam’s CI][CI]:

    c:\projects\opam\bootstrap\ocaml\lib\ocaml\caml\domain_state.h(49): error C2118: negative subscript

It comes from here:

https://github.com/ocaml/ocaml/blob/b212928e740388d55ebce1978541e02db77cbb20/runtime/caml/domain_state.h#L47-L50

If the assertion fails, it emits an array of size -1 which the
compiler rejects.

https://github.com/ocaml/ocaml/blob/b212928e740388d55ebce1978541e02db77cbb20/runtime/caml/domain_state.h#L28-L36

The structure is carefully crafted so that each of its elements are
aligned to an 8-bit boundary, meaning that no padding is inserted
*between* the elements.

https://github.com/ocaml/ocaml/blob/b212928e740388d55ebce1978541e02db77cbb20/runtime/caml/misc.h#L93-L105

The `CAMLalign` is well defined to support most compilers; so we can
guess that all elements are aligned as expected, and no padding is
inserted *between* them.

However, this fails to take into account the padding that the compiler
may insert *at the end* of the structure. First mistake. Ref: [The Lost Art of Structure Packing, §4](http://www.catb.org/esr/structure-packing/#_padding). 


https://github.com/ocaml/ocaml/blob/b212928e740388d55ebce1978541e02db77cbb20/runtime/caml/domain_state.h#L38-L43

`Domain_state_num_fields` counts the number of fields in the table.
The comparison with `Domain_state_num_fields * 8` can not be correct
if some fields are not 8 octects long. One only need to see that the
`domain_state.tbl` stores pointers, and they are not always 8
octets-long, especially on 32-bits architectures. Second mistake.


Now, to fix the assertion, we could loop over the fields, and keep
accumulate the sum of the sizes of each fields, and compare that with
the `caml_domain_state` structure, but again, we would miss the end
structure padding.

I think the best fix is to *pack* and *align* to the 8-bit boundary
the structure.
EDIT: and we could remove calls to `CAMLalign` and the static assertion.

So, with [GCC][GCC-attributes] we’re interested by the `packed`, `aligned`, and
`warn_if_not_aligned` attributes, but that wouldn’t be portable.
An alternative with pragmas that seems to be compatible across
compilers: [GCC][GCC-pragmas], [Clang][Clang-pragmas] (manpage is old but the feature should
still be present), and [MSVC][MSVC-pragmas].

```c
#pragma pack(push, 8)
typedef struct {
    // …
} caml_domain_state;
#pragma pack(pop)
```

https://github.com/ocaml/ocaml/blob/b212928e740388d55ebce1978541e02db77cbb20/runtime/caml/misc.h#L93-L96

This statement puzzles me. Could the runtime use `offsetof` to get the
address of each member?

I hope I got the analysis right. What do you think?

-- Antonin

[41000-bootstrap]: https://github.com/ocaml/opam/pull/4242
[CI]: https://ci.appveyor.com/project/AltGr/opam/build/job/qenst562ctllutth#L6672
[GCC-attributes]: https://gcc.gnu.org/onlinedocs/gcc/Common-Type-Attributes.html#Common-Type-Attributes
[GCC-pragmas]: https://gcc.gnu.org/onlinedocs/gcc/Structure-Layout-Pragmas.html#Structure-Layout-Pragmas
[Clang-pragmas]: https://releases.llvm.org/3.4/tools/clang/docs/UsersManual.html#microsoft-extensions
[MSVC-pragmas]: https://docs.microsoft.com/en-us/cpp/preprocessor/pack?view=vs-2019

--- a/Changes
+++ b/Changes
@@ -257,6 +257,12 @@ Working version
   (Xavier Leroy, Sadiq Jaffer, Gabriel Scherer,
    review by Xavier Leroy and Jacques-Henri Jourdan)
 
+- #9714, #9724: Use the C++ alignas keyword when compiling in C++.
+  Fixes a bug with MSVC C++ 2015/2017. Add a terminator to the
+  `caml_domain_state` structure to better ensure that members are
+  correctly spaced.
+  (Antonin Décimo, review by David Allsopp and Xavier Leroy)
+
 OCaml 4.11
 ----------
 

--- a/runtime/caml/domain_state.h
+++ b/runtime/caml/domain_state.h
@@ -33,6 +33,7 @@ typedef struct {
 #endif
 #include "domain_state.tbl"
 #undef DOMAIN_STATE
+    CAMLalign(8) char end_of_domain_state;
 } caml_domain_state;
 
 enum {
@@ -45,9 +46,8 @@ enum {
 /* Check that the structure was laid out without padding,
    since the runtime assumes this in computing offsets */
 CAML_STATIC_ASSERT(
-  sizeof(caml_domain_state) ==
-   (Domain_state_num_fields
-   ) * 8);
+    offsetof(caml_domain_state, end_of_domain_state) ==
+    Domain_state_num_fields * 8);
 
 CAMLextern caml_domain_state* Caml_state;
 #ifdef CAML_NAME_SPACE

--- a/runtime/caml/misc.h
+++ b/runtime/caml/misc.h
@@ -96,6 +96,8 @@ CAMLdeprecated_typedef(addr, char *);
 /* we need to be able to compute the exact offset of each member. */
 #if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
 #define CAMLalign(n) _Alignas(n)
+#elif defined(__cplusplus) && (__cplusplus >= 201103L || _MSC_VER >= 1900)
+#define CAMLalign(n) alignas(n)
 #elif defined(SUPPORTS_ALIGNED_ATTRIBUTE)
 #define CAMLalign(n) __attribute__((aligned(n)))
 #elif _MSC_VER >= 1500

